### ISSUE TITLE
Optionally save version properties

### ISSUE THREAD

**Issue by lewismorgan** [ethauvin/semver-gradle#1]
I'm currently working on refactoring a fairly large application to use some semver standards and came across this gradle plugin. The rest I could find were massive and I just needed something small and simple, and this has been working good thus far.

Currently, properties are saved every single time the [plugin is applied to a project](https://github.com/ethauvin/semver-gradle/blob/master/src/main/kotlin/net/thauvin/erik/gradle/semver/SemverPlugin.kt#L89)  [since its located in afterEvaluate](https://github.com/ethauvin/semver-gradle/blob/master/src/main/kotlin/net/thauvin/erik/gradle/semver/SemverPlugin.kt#L135)

This is causing a problem for me since I only want to create build meta's on distribution tasks. However, version.properties is getting saved every time during any gradle task being performed, due to afterEvaluate. It wouldn't really be a problem but Properties applies a timestamp so version.properties is showing up in the changelog 24/7 even though I haven't incremented anything.

I'd like to add a configuration option to disable the saving at that line, unless there is a reason right now it's intentionally doing a properties save every project evaluation.

**Comment by ethauvin** [comment]
@lewismorgan That's a very good point.

As far as I can remember, the only reason the properties file is saved after evaluation is to ensure that the default version properties are set if missing (first run, etc.)

There should be no reason to overwrite the properties file if the version properties are already set, which would work for you. That's a little bit more work than a configuration option to disable the saving altogether, but it makes more sense to me.

I don't know if you want to take a crack at it. I would but I just had major surgery on Friday and I'm in no shape to do it right now. Let me know.

**Comment by lewismorgan** [comment]
I did a bit looking into it and it seems like it'll be doable. I noticed that the kotlin dsl isn't setup fully yet, but I did get it working by just adding a semver and then having the config as a object. Then in the build script it's just simply doing this:

```
configure<SemverPluginExtension> {
    config.saveOnApply = false // Enabled by default
}
```

The main changes I had to do were creating the new class `SemverPluginExtension`, which just serves as a wrapper for gradle's ObjectFactory injection with a SemverConfig class. As a result, the DSL will end up looking like below...
```
semver {
    config {
        prereleasePrefixKey = "SNAPSHOT"
        saveOnApply = false // true by default to keep functionality the same as it is right now
    }
}
```

Or it can be changed to just having everything within the extension. I think the variation below will be the best and prevents confusion. I'll just need to move the ObjectFactory usage into the SemverConfig as a constructor, which now with what I know about Gradle plugins won't be much of a time consuming process:
```
semver {
    prereleasePrefixKey = "SNAPSHOT"
    saveOnApply = false
}
```

Now for the saving to actually work (or not, heh), I just did the following:
```
if (config.saveOnApply) {
    saveProperties(config, version)
}
```

I was worried that this would lead to issues with saving in general, but it appears not to be an issue. The file is saved whenever the `increment` method is called already. There was also a slight change to the ordering of code. Extension functions weren't working and it was driving me mad, it was because they need to be added before the afterEvaluate call:
```
val extension = project.extensions.create<SemverPluginExtension>("semver", project.objects)
config = extension.config
project.afterEvaluate(this::afterEvaluate)
```

As far as my project goes, the task is super simple now as I can tell gradle to just go ahead and throw in a hash for just the distributions (mainly CI builds). Now, the only time version.properties pops up in the change log is a major/minor/patch update which is exactly what I was looking for :)

tl;dr - Let me know what you think about the DSL and I can get a PR in on it.

Best wishes to you on a fast and speedy recovery!

**Comment by ethauvin** [comment]
@lewismorgan Can you do me a favor, and try your modification with a project that has no `version.properties` file and no version increment in the build file? Does it work?

Thanks.

**Comment by lewismorgan** [comment]
@ethauvin That was actually a problem when I was testing earlier a couple days ago working on it. This actually was not working and throwing a gradle exception because File.canWrite() will always return false if the file does not exist. I have it accounted for in the PR I submitted a few minutes ago.

I did however account that by not saving the file and having the new config option added would never create it, but that was a simple check to see if the file exists or not. If the props file doesn't exist then the config option is ignored until the next project evaluation.

See PR https://github.com/ethauvin/semver-gradle/pull/2

**Comment by ethauvin** [comment]
@lewismorgan Great! I'll have a look at it as soon as I can. Thanks.

**Comment by ethauvin** [comment]
@lewismorgan I've merged your PR and made a few modifications that should make it so that your `saveAfterProjectEvaluate` option is no longer needed. Bascially, I'm checking to see wether the required version properties are in the file already, if not, it'll be saved.

I've lightly tested it and it seems to work for me. Give it a shot and let me know if that works for your project(s). If it does, I'll push a new version out.

Thanks!

**Comment by lewismorgan** [comment]
Working for me!

**Comment by ethauvin** [comment]
Fixed in https://github.com/ethauvin/semver-gradle/commit/2baf69cd32891952ae32dbea8e223cc4592b56a3.
